File: LeetCode/Remove Element/cpp/remove_element.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    public:
    int removeElement(vector<int>& arr, int val) {
        
        int i = 0;
        for(size_t j = 0; j < arr.size(); j++) {
            if (arr[j] != val) {
                arr[i] = arr[j];
                i++;
            }
        }

        return i;
    }

    int removeElementOptimize(vector<int>& arr, int val) {
        
        int i = 0;
        int n = arr.size();

        while (i < n) {
            if(arr[i] == val) {
                arr[i] = arr[n-1];
                n--;
            } else {
                i++;
            }

        }

        return n;
    }
    
};

int main() {

    vector<int> nums = {3,2,2,3};
    int val = 3;

    Solution solution;

    cout << solution.removeElement(nums, val) << endl;
    
    nums = {3,2,2,3};
    cout << solution.removeElementOptimize(nums, val) << endl;

    return 0;
}

File: LeetCode/Remove Element/java/src/RemoveElement.java
public class RemoveElement {

    public static int removeElement(int[] nums, int val) {
        int i = 0;
        for (int j = 0; j < nums.length; j++) {
            if (nums[j] != val) {
                nums[i] = nums[j];
                i++;
            }
        }
        return i;
    }

    public static int removeElementOptimize(int[] nums, int val) {
        int i = 0;
        int n = nums.length;
        while (i < n) {
            if (nums[i] == val) {
                nums[i] = nums[n - 1];
                // reduce array size by one
                n--;
            } else {
                i++;
            }
        }
        return n;
    }

    public static void main(String[] args) {

        int[] nums1 = {3,2,2,3};
        int val = 3;

        System.out.println(removeElement(nums1, val));

        int[] nums2 = {3,2,2,3};
        System.out.println(removeElementOptimize(nums2, val));
    }
}
